graphix/sim/statevec.py

```diff
@@ -178,28 +202,102 @@ def meas_op(angle, vop=0, plane="XY", choice=0):
+    def __init__(
+        self,
+        data: typing.Optional[SV_Data] = graphix.states.BasicStates.PLUS,
+        nqubit: typing.Optional[graphix.types.PositiveOrNullInt] = None,
+    ):
+        """Initialize statevector objects. The behaviour is as follows. `data` can be:
+        - a single :class:`graphix.states.State` (classical description of a quantum state)
+        - an iterable of :class:`graphix.states.State` objects
+        - an iterable of scalars (A 2**n numerical statevector)
+        - a `graphix.statevec.Statevec` object
+
+        If `nqubit` is not provided, the number of qubit is inferred from `data` and checked for consistency.
+        If only one :class:`graphix.states.State` is provided and nqubit is a valid integer, initialize the statevector
+        in the tensor product state.
+        If both `nqubit` and `data` are provided, consistency of the dimensions is checked.
+        If a `graphix.statevec.Statevec` is passed, returns a copy.
+
+
+        :param data: input data to prepare the state. Can be a classical description or a numerical input, defaults to graphix.states.BasicStates.PLUS
+        :type data: typing.Union[
+        graphix.states.State, "Statevec", typing.Iterable[graphix.states.State], typing.Iterable[numbers.Number]
+        ], optional
+        :param nqubit: number of qubits to prepare, defaults to None
+        :type nqubit: int, optional
+        """
 
-    def __init__(self, nqubit=1, plus_states=True):
-        """Initialize statevector
+        assert nqubit is None or isinstance(nqubit, numbers.Integral) and nqubit >= 0
 
-        Parameters
-        ----------
-        nqubit : int, optional:
-            number of qubits. Defaults to 1.
-        plus_states : bool, optional
-            whether or not to start all qubits in + state or 0 state. Defaults to +
-        """
-        if plus_states:
-            self.psi = np.ones((2,) * nqubit) / 2 ** (nqubit / 2)
+        if isinstance(data, Statevec):
+            # assert nqubit is None or len(state.flatten()) == 2**nqubit
+            if nqubit is not None and len(data.flatten()) != 2**nqubit:
+                raise ValueError(
+                    f"Inconsistent parameters between nqubit = {nqubit} and the inferred number of qubit = {len(data.flatten())}."
+                )
+            self.psi = data.psi.copy()
+            return
+
+        if isinstance(data, graphix.states.State):
+            if nqubit is None:
+                nqubit = 1
+            input_list = [data] * nqubit
+        elif isinstance(data, typing.Iterable):
+            input_list = list(data)
+        else:
+            raise TypeError(f"Incorrect type for data: {type(data)}")
+
+        if len(input_list) == 0:
+            if nqubit is not None and nqubit != 0:
+                raise ValueError("nqubit is not null but input state is empty.")
+
+            # warnings.warn(f"Called Statevec with 0 qubits. Ignoring the state.")
```

Comment:
are we not keeping this warning?

Reply:
No. That is inherited from a previous version. Now if there is a mismatch between the dimensions, an Exception will be raised. The only acceptable behaviour for `nqubit = 0` is without an input state as used in the `add_nodes()` method.